language-plutus-core/test/Evaluation/CkMachine.hs

```diff
@@ -0,0 +1,115 @@
+{-# LANGUAGE OverloadedStrings #-}
+module Evaluation.CkMachine
+    ( test_evaluateCk
+    ) where
+
+import           PlutusPrelude hiding (hoist, list)
+import           Language.PlutusCore
+import           Language.PlutusCore.Constant
+import           Language.PlutusCore.CkMachine
+import           Language.PlutusCore.TestSupport
+import           Language.PlutusCore.StdLib.Data.Bool
+import           Language.PlutusCore.StdLib.Data.Function
+import           Language.PlutusCore.StdLib.Data.List
+import           Language.PlutusCore.StdLib.Data.Nat
+import           Language.PlutusCore.StdLib.Data.Unit
+
+import           Data.Foldable
+import           Control.Monad.Morph
+import           Hedgehog hiding (Size, Var, annotate)
+import qualified Hedgehog.Gen as Gen
+import qualified Hedgehog.Range as Range
+import           Test.Tasty
+import           Test.Tasty.Hedgehog
+
+-- | Convert an 'Integer' to a @nat@.
+getBuiltinIntegerToNat :: Integer -> Quote (Term TyName Name ())
+getBuiltinIntegerToNat n
+    | n < 0     = error $ "getBuiltinIntegerToNat: negative argument: " ++ show n
+    | otherwise = go n where
+          go 0 = getBuiltinZero
+          go m = Apply () <$> getBuiltinSucc <*> go (m - 1)
+
+-- | Convert a @nat@ to an 'Integer'.
+getBuiltinNatToInteger :: Natural -> Term TyName Name () -> Quote (Term TyName Name ())
+getBuiltinNatToInteger s n = do
+    builtinFoldNat <- getBuiltinFoldNat
+    let int = Constant () . BuiltinInt () s
+    return
+        . foldl (Apply ()) (TyInst () builtinFoldNat $ TyBuiltin () TyInteger)
+        $ [ Apply () (Constant () $ BuiltinName () AddInteger) $ int 1
+          , int 0
+          , n
+          ]
+
+-- | Generate an 'Integer', turn it into a Scott-encoded PLC @Nat@ (see 'getBuiltinNat'),
+-- turn that @Nat@ into the corresponding PLC @integer@ using a fold (see 'getBuiltinFoldNat')
+-- defined in terms of generic fix (see 'getBuiltinFix'), feed the resulting 'Term' to the CK machine
+-- (see 'evaluateCk') and check that the original 'Integer' and the computed @integer@ are in sync.
+test_NatRoundtrip :: TestTree
+test_NatRoundtrip = testProperty "NatRoundtrip" . property $ do
+    let size = 1
+        int1 = TypedBuiltinSized (SizeValue size) TypedBuiltinSizedInt
+    TermOf n nv <- forAllPrettyCfg . Gen.filter ((>= 0) . _termOfValue) $ genTypedBuiltinDef int1
+    let term = runQuote $ getBuiltinIntegerToNat nv >>= getBuiltinNatToInteger size
+    evaluateCk term === CkEvalSuccess n
```

Comment:
Should we be typechecking the terms first? That would give us typechecker coverage too.

Reply:
Yes, but last time I tried to type check things, I got errors. And these are actually working tests. Also we have type checker coverage in a [separate module](https://github.com/input-output-hk/plutus-prototype/blob/ba4616425fb3f9b259985c8e91d41578785396e4/language-plutus-core/test/Evaluation/TypeCheck.hs).

Reply:
I hope you're reporting the bugs to Vanessa :)

Reply:
Of course :)

Reply:
Done.